src/Microsoft.Identity.Client/Azure/ManagedIdentity.cs

```diff
@@ -0,0 +1,370 @@
+                var provider = BuildProvider(maxRetries: 3, httpClient: _httpClient);
+                var token = await provider.GetTokenAsync(new List<string> { @"https://management.azure.com//.default" }).ConfigureAwait(false);
+                return token != null;
+            }
+            catch (MsalException ex)
+            {
+                // If we tried, but were unable to fetch a token, then false
+                if (ex.ErrorCode != ErrorCodes.MaxRetriesExhausted)
+                {
+                    throw ex;
+                }
+                return false;
+            }
+        }
+
+        /// <summary>
+        /// IsAppService tells us if we are executing within AppService with Managed Identities enabled
+        /// </summary>
+        /// <returns></returns>
+        private bool IsAppService()
+        {
+            var vars = new List<string> { _config.ManagedIdentitySecret, _config.ManagedIdentityEndpoint };
+            return vars.All(item => !string.IsNullOrWhiteSpace(item));
+        }
+
+#if net45 || NETSTANDARD2_0
+        private static bool IsVMManagedIdentityListening()
+        {
+            var loopback = IPAddress.Parse(Constants.ManagedIdentityLoopbackAddress);
+            foreach (var tcpEndpoint in IPGlobalProperties.GetIPGlobalProperties().GetActiveTcpListeners())
+            {
+                if (tcpEndpoint.Address == loopback && tcpEndpoint.Port == 80)
+                {
+                    return true;
+                }
+            }
+            return false;
+        }
+#endif
+
+        /// <summary>
+        ///     Create a managed identity credential provider from the information discovered by the probe
+        /// </summary>
+        /// <returns>A managed identity credential provider instance</returns>
+        public async Task<ITokenProvider> ProviderAsync()
+        {
+            if (!await AvailableAsync().ConfigureAwait(false))
+                throw new InvalidOperationException("The required environment variables are not available.");
+
+            return BuildProvider(httpClient: _httpClient);
+        }
+
+        private ITokenProvider BuildProvider(int maxRetries = 5, HttpClient httpClient = null)
+        {
+            var endpoint = IsAppService() ? _config.ManagedIdentityEndpoint : Constants.ManagedIdentityTokenEndpoint;
+            return new ManagedIdentityCredentialProvider(endpoint, httpClient: httpClient, secret: _config.ManagedIdentitySecret, clientId: ClientId, maxRetries: maxRetries);
+        }
+
+        private string ClientId
+        {
+            get { return string.IsNullOrWhiteSpace(_overrideClientId) ? _config.ClientId : _overrideClientId; }
+        }
+    }
+
+    /// <summary>
+    ///     ManagedIdentityCredentialProvider will fetch AAD JWT tokens from the IMDS endpoint for the default client id or
+    ///     a specified client id.
+    /// </summary>
+    public class ManagedIdentityCredentialProvider : ITokenProvider
```

Comment:
Should it be public?

Reply:
The intention is to allow developers to build their own chain of IProbes and ITokenProviders. These are intended to be public and suitable for extension.